### PR TITLE
MAVLink: reply to parameter requests asynchronously

### PR DESCRIPTION
**Pull request by tridge**
This adds an IO timer in GCS_MAVLink to reply to parameter request read requests asynchronously. This prevents a denial of service attack by a GCS that requests many PARAM_REQUEST_READ requests rapidly, with each one triggering a full parameter tree scan.

### CODE REVIEW

File: libraries/AP_Param/AP_Param.cpp [hunk 1]
```diff
@@ -1989,16 +1989,20 @@ void AP_Param::send_parameter(const char *name, enum ap_var_type var_type, uint8
...
 {
     // if we haven't cached the parameter count yet...
-    if (0 == _parameter_count) {
+    uint16_t ret = _parameter_count;
+    if (0 == ret) {
         AP_Param  *vp;
         AP_Param::ParamToken token;
 
         vp = AP_Param::first(&token, nullptr);
+        uint16_t count = 0;
```
Comment (peterbarker) [inline libraries/AP_Param/AP_Param.cpp]: I assume this is for thread safety.  A comment along those lines would seem appropriate.

File: libraries/AP_Param/AP_Param.cpp [hunk 2]
```diff
@@ -1989,16 +1989,20 @@ void AP_Param::send_parameter(const char *name, enum ap_var_type var_type, uint8
 uint16_t AP_Param::count_parameters(void)
 {
     // if we haven't cached the parameter count yet...
-    if (0 == _parameter_count) {
+    uint16_t ret = _parameter_count;
```
Comment (peterbarker) [inline libraries/AP_Param/AP_Param.cpp]: Is a temp for this required?

File: libraries/GCS_MAVLink/GCS_Param.cpp [hunk 3]
```diff
@@ -200,38 +218,23 @@ void GCS_MAVLINK::handle_param_request_read(mavlink_message_t *msg)
...
-        msg,
-        chan,
-        param_name,
-        value,
-        mav_var_type(p_type),
-        AP_Param::count_parameters(),
-        packet.param_index);
 }
-
+
```
Comment (peterbarker) [inline libraries/GCS_MAVLink/GCS_Param.cpp]: Whitespace :-)

Comment (peterbarker) [inline libraries/GCS_MAVLink/GCS_Param.cpp]:
```
(add-hook 'c++-mode-hook (lambda () (setq show-trailing-whitespace 1)))
```

File: libraries/GCS_MAVLink/GCS_Param.cpp [hunk 4]
```diff
@@ -359,3 +363,70 @@ bool GCS_MAVLINK::send_queued_parameters(void)
...
+    if (req.param_index != -1) {
+        AP_Param::ParamToken token;
+        vp = AP_Param::find_by_index(req.param_index, &reply.p_type, &token);
+        if (vp == nullptr) {
+            return;
+        }
+        vp->copy_name_token(token, reply.param_name, AP_MAX_NAME_SIZE, true);
+    } else {
+        strncpy(reply.param_name, req.param_name, AP_MAX_NAME_SIZE+1);
```
Comment (peterbarker) [inline libraries/GCS_MAVLink/GCS_Param.cpp]: We should standardise on memcpy or strncpy for these copies.

Comment (tridge) [inline libraries/GCS_MAVLink/GCS_Param.cpp]: the reason for using memcpy in that one place is a valgrind false alarm

File: libraries/AP_Param/AP_Param.cpp (hunk 2 above)
Comment (tridge) [inline libraries/AP_Param/AP_Param.cpp]: fixed

**Comment by tridge** [comment]
merged
thanks for the review to @peterbarker @OXINARF

**Pull request closed without merging**